`package-manager-strict` enforcement in pnpm 9 is a massive headache

### Last pnpm version that worked

8

### pnpm version

9

### Code to reproduce the issue

`pnpm install` with `packageManager` set to a different minor or patch version than you have installed

### Expected behavior

I would expect to use the `engines` field to manage restrictions on what versions of package managers.

The `packageManager` field is used by platforms like Vercel and Netlify to choose what exact version of pnpm to use to run your project, which has been helpful for ensuring they're using a compatible major version.

However, I do not want to have to sync all of my various projects and locally installed version to always be using the same exact version of pnpm - especially with how frequently pnpm updates are released. Now I will have to constantly be updating the `packageManager` field in all my projects to match whatever the latest pnpm version is causing needless churn and a ton of annoying work that provides close to zero value. If I wanted to do that I already had the ability via `engines`.

Other users have been frustrated by this new enforcement as well. E.g. see https://github.com/pnpm/pnpm/issues/7956#issuecomment-2063238126

### Actual behavior

> ERR_PNPM_BAD_PM_VERSION  This project is configured to use v9.1.1 of pnpm. Your current pnpm is v9.1.0


### Additional information

Support for semver ranges is the number two issue in corepack https://github.com/nodejs/corepack/issues/95. Setting `package-manager-strict` could make more sense after that is available. Though it would still seem to unnecessarily duplicate the `engines` field, so even then it wouldn't be my preference

### Node.js version

20.10.0

### Operating System

Linux

I just ran into this issue on Vercel after I had been deploying for days without issue? `postinstall:  ERR_PNPM_BAD_PM_VERSION  This project is configured to use v9.1.0 of pnpm. Your current pnpm is v9.0.4`. Did Vercel all of a sudden stop respecting `"packageManager": "pnpm@9.1.0"`? Curious if anyone else started running into this.

> I just ran into this issue on Vercel after I had been deploying for days without issue? `postinstall:  ERR_PNPM_BAD_PM_VERSION  This project is configured to use v9.1.0 of pnpm. Your current pnpm is v9.0.4`. Did Vercel all of a sudden stop respecting `"packageManager": "pnpm@9.1.0"`? Curious if anyone else started running into this.

I have the same issue.
Currently we use in CI

```bash
corepack use pnpm@`pnpm -v`
```
as a work around. 

![image](https://github.com/pnpm/pnpm/assets/1060633/52fcb1d8-74cc-45f8-9e19-ce6d2aab7a6f)

--- a/.changeset/unlucky-avocados-talk.md
+++ b/.changeset/unlucky-avocados-talk.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/cli-utils": patch
+"pnpm": patch
+---
+
+pnpm doesn't fail if its version doesn't match the one specified in the "packageManager" field of `package.json` [#8087](https://github.com/pnpm/pnpm/issues/8087).

--- a/.changeset/young-carpets-stare.md
+++ b/.changeset/young-carpets-stare.md
@@ -1,0 +1,8 @@
+---
+"@pnpm/workspace.find-packages": minor
+"@pnpm/cli-utils": minor
+"@pnpm/config": minor
+"pnpm": minor
+---
+
+If `package-manager-strict-version` is set to `true` pnpm will fail if its version will not exactly match the version in the `packageManager` field of `package.json`.

--- a/cli/cli-utils/src/packageIsInstallable.ts
+++ b/cli/cli-utils/src/packageIsInstallable.ts
@@ -15,16 +15,17 @@ export function packageIsInstallable (
   },
   opts: {
     packageManagerStrict?: boolean
+    packageManagerStrictVersion?: boolean
     engineStrict?: boolean
     nodeVersion?: string
     supportedArchitectures?: SupportedArchitectures
   }
 ): void {
-  const pnpmVersion = packageManager.name === 'pnpm'
+  const currentPnpmVersion = packageManager.name === 'pnpm'
     ? packageManager.version
     : undefined
   if (pkg.packageManager && !process.env.COREPACK_ROOT) {
-    const [pmName] = pkg.packageManager.split('@')
+    const [pmName, pmReference] = pkg.packageManager.split('@')
     if (pmName && pmName !== 'pnpm') {
       const msg = `This project is configured to use ${pmName}`
       if (opts.packageManagerStrict) {
@@ -32,11 +33,24 @@ export function packageIsInstallable (
       } else {
         globalWarn(msg)
       }
+    } else if (opts.packageManagerStrictVersion && !pmReference.includes(':')) {
+      // pmReference is semantic versioning, not URL
+      const [requiredPnpmVersion] = pmReference.split('+')
+      if (requiredPnpmVersion && currentPnpmVersion && requiredPnpmVersion !== currentPnpmVersion) {
+        const msg = `This project is configured to use v${requiredPnpmVersion} of pnpm. Your current pnpm is v${currentPnpmVersion}`
+        if (opts.packageManagerStrict) {
+          throw new PnpmError('BAD_PM_VERSION', msg, {
+            hint: 'If you want to bypass this version check, you can set the "package-manager-strict" configuration to "false" or set the "COREPACK_ENABLE_STRICT" environment variable to "0"',
+          })
+        } else {
+          globalWarn(msg)
+        }
+      }
     }
   }
   const err = checkPackage(pkgPath, pkg, {
     nodeVersion: opts.nodeVersion,
-    pnpmVersion,
+    pnpmVersion: currentPnpmVersion,
     supportedArchitectures: opts.supportedArchitectures ?? {
       os: ['current'],
       cpu: ['current'],

--- a/cli/cli-utils/src/packageIsInstallable.ts
+++ b/cli/cli-utils/src/packageIsInstallable.ts
@@ -24,26 +24,13 @@ export function packageIsInstallable (
     ? packageManager.version
     : undefined
   if (pkg.packageManager && !process.env.COREPACK_ROOT) {
-    const [pmName, pmReference] = pkg.packageManager.split('@')
+    const [pmName] = pkg.packageManager.split('@')
     if (pmName && pmName !== 'pnpm') {
       const msg = `This project is configured to use ${pmName}`
       if (opts.packageManagerStrict) {
         throw new PnpmError('OTHER_PM_EXPECTED', msg)
       } else {
         globalWarn(msg)
-      }
-    } else if (!pmReference.includes(':')) {
-      // pmReference is semantic versioning, not URL
-      const [pmVersion] = pmReference.split('+')
-      if (pmVersion && pnpmVersion && pmVersion !== pnpmVersion) {
-        const msg = `This project is configured to use v${pmVersion} of pnpm. Your current pnpm is v${pnpmVersion}`
-        if (opts.packageManagerStrict) {
-          throw new PnpmError('BAD_PM_VERSION', msg, {
-            hint: 'If you want to bypass this version check, you can set the "package-manager-strict" configuration to "false" or set the "COREPACK_ENABLE_STRICT" environment variable to "0"',
-          })
-        } else {
-          globalWarn(msg)
-        }
       }
     }
   }

--- a/cli/cli-utils/src/readProjectManifest.ts
+++ b/cli/cli-utils/src/readProjectManifest.ts
@@ -5,6 +5,7 @@ import { packageIsInstallable } from './packageIsInstallable'
 export interface ReadProjectManifestOpts {
   engineStrict?: boolean
   packageManagerStrict?: boolean
+  packageManagerStrictVersion?: boolean
   nodeVersion?: string
   supportedArchitectures?: SupportedArchitectures
 }

--- a/config/config/src/Config.ts
+++ b/config/config/src/Config.ts
@@ -192,6 +192,7 @@ export interface Config {
   dedupeInjectedDeps?: boolean
   nodeOptions?: string
   packageManagerStrict?: boolean
+  packageManagerStrictVersion?: boolean
   virtualStoreDirMaxLength: number
 }
 

--- a/config/config/src/index.ts
+++ b/config/config/src/index.ts
@@ -245,6 +245,7 @@ export async function getConfig (
     'package-lock': npmDefaults['package-lock'],
     pending: false,
     'package-manager-strict': process.env.COREPACK_ENABLE_STRICT !== '0',
+    'package-manager-strict-version': false,
     'prefer-workspace-packages': false,
     'public-hoist-pattern': [
       '*eslint*',

--- a/pnpm/test/install/misc.ts
+++ b/pnpm/test/install/misc.ts
@@ -267,9 +267,12 @@ test('install should not fail if the used pnpm version does not satisfy the pnpm
     packageManager: 'pnpm@0.0.0',
   })
 
-  const { status } = execPnpmSync(['install'])
+  expect(execPnpmSync(['install']).status).toBe(0)
 
-  expect(status).toBe(0)
+  const { status, stdout } = execPnpmSync(['install', '--config.package-manager-strict-version=true'])
+
+  expect(status).toBe(1)
+  expect(stdout.toString()).toContain('This project is configured to use v0.0.0 of pnpm. Your current pnpm is')
 })
 
 test('install should fail if the project requires a different package manager', async () => {

--- a/pnpm/test/install/misc.ts
+++ b/pnpm/test/install/misc.ts
@@ -259,7 +259,7 @@ test('install should fail if the used pnpm version does not satisfy the pnpm ver
   expect(stdout.toString()).toContain('Your pnpm version is incompatible with')
 })
 
-test('install should fail if the used pnpm version does not satisfy the pnpm version specified in packageManager', async () => {
+test('install should not fail if the used pnpm version does not satisfy the pnpm version specified in packageManager', async () => {
   prepare({
     name: 'project',
     version: '1.0.0',
@@ -267,17 +267,9 @@ test('install should fail if the used pnpm version does not satisfy the pnpm ver
     packageManager: 'pnpm@0.0.0',
   })
 
-  const { status, stdout } = execPnpmSync(['install'])
+  const { status } = execPnpmSync(['install'])
 
-  expect(status).toBe(1)
-  expect(stdout.toString()).toContain('This project is configured to use v0.0.0 of pnpm. Your current pnpm is')
-
-  expect(execPnpmSync(['install', '--config.package-manager-strict=false']).status).toBe(0)
-  expect(execPnpmSync(['install'], {
-    env: {
-      COREPACK_ENABLE_STRICT: '0',
-    },
-  }).status).toBe(0)
+  expect(status).toBe(0)
 })
 
 test('install should fail if the project requires a different package manager', async () => {

--- a/workspace/find-packages/src/index.ts
+++ b/workspace/find-packages/src/index.ts
@@ -12,6 +12,7 @@ export async function findWorkspacePackages (
   opts?: {
     engineStrict?: boolean
     packageManagerStrict?: boolean
+    packageManagerStrictVersion?: boolean
     nodeVersion?: string
     patterns?: string[]
     sharedWorkspaceLockfile?: boolean

